core/src/main/kotlin/entity/interaction/OptionValue.kt

```diff
@@ -0,0 +1,102 @@
+package dev.kord.core.entity.interaction
+
+import dev.kord.common.entity.CommandArgument
+import dev.kord.common.entity.Snowflake
+import dev.kord.core.behavior.interaction.GlobalInteractionBehavior
+import dev.kord.core.entity.*
+import dev.kord.core.entity.channel.ResolvedChannel
+
+public sealed interface OptionValue<out T> {
+    public val value: T
+    public val focused: Boolean
+}
+
+public sealed interface ResolvableOptionValue<out T> : OptionValue<Snowflake> {
+    public val resolvedObject: T?
+}
+
+public class IntOptionValue(override val value: Long, override val focused: Boolean) : OptionValue<Long> {
+    override fun toString(): String = "IntOptionValue(value=$value)"
+}
+
+
+public class NumberOptionValue(override val value: Double, override val focused: Boolean) : OptionValue<Double> {
+    override fun toString(): String = "DoubleOptionValue(value=$value)"
+}
+
+public class StringOptionValue(override val value: String, override val focused: Boolean) : OptionValue<String> {
+    override fun toString(): String = "StringOptionValue(value=$value)"
+}
+
+public class BooleanOptionValue(override val value: Boolean, override val focused: Boolean) : OptionValue<Boolean> {
+    override fun toString(): String = "BooleanOptionValue(value=$value)"
+}
+
+
+public class RoleOptionValue(override val value: Snowflake, override val focused: Boolean, override val resolvedObject: Role?) : ResolvableOptionValue<Role> {
+    override fun toString(): String = "RoleOptionValue(value=$value)"
+}
+
+public open class UserOptionValue(override val value: Snowflake, override val focused: Boolean, override val resolvedObject: User?) : ResolvableOptionValue<User> {
+    override fun toString(): String = "UserOptionValue(value=$value)"
+}
+
+public class MemberOptionValue(value: Snowflake, focused: Boolean, override val resolvedObject: Member?) : UserOptionValue(value, focused, resolvedObject) {
+    override fun toString(): String = "MemberOptionValue(value=$value)"
+}
+
+public class ChannelOptionValue(override val value: Snowflake, override val focused: Boolean, override val resolvedObject: ResolvedChannel?) :
+    ResolvableOptionValue<ResolvedChannel> {
+    override fun toString(): String = "ChannelOptionValue(value=$value)"
+}
+public class MentionableOptionValue(override val value: Snowflake, override val focused: Boolean, override val resolvedObject: Entity?) : ResolvableOptionValue<Entity> {
+    override fun toString(): String = "MentionableOptionValue(value=$value)"
+}
```

Comment:
Entity is the lowest; we are insterested in keeping it an Entity instead of a Kord Entity

Reply:
I thought it could be useful since it is in an `out` position so we can specify the most concrete type. But I can see that there is not much of a value in having the `kord` property on it, it's in the interaction and event anyway.